python/ray/data/_internal/execution/backpressure_policy/streaming_output_backpressure_policy.py

```diff
@@ -59,28 +70,79 @@ def __init__(self, topology: "Topology"):
-                # deadlock check more accurate by directly checking resources.
+
+            if downstream_idle:
                 max_blocks_to_read_per_op[state] = max(
                     max_blocks_to_read_per_op[state],
                     1,
                 )
-            downstream_num_active_tasks += len(op.get_active_tasks())
+
+            # An operator is considered idle if either of the following is true:
+            # - It has no active tasks.
+            #   - This can happen when all resources are used by upstream operators.
+            # - It has active tasks, but no outputs for at least
+            #   `MAX_OUTPUT_IDLE_SECONDS`.
+            #   - This can happen when non-Data code preempted cluster resources, and
+            #   - some of the active tasks don't actually have enough resources to run.
+            #
+            # If the operator is idle, we'll temporarily unblock backpressure by
+            # allowing reading at least one block from its upstream
+            # to avoid deadlock.
+            # NOTE, these 2 conditions don't necessarily mean deadlock.
+            # The first case can also happen when the upstream operator hasn't outputted
+            # any blocks yet. While the second case can also happen when the task is
+            # expected to output data slowly.
+            # The false postive cases are fine as we only allow reading one block
+            # each time.
+            downstream_idle = False
+            if op.num_active_tasks() == 0:
+                downstream_idle = True
+            else:
+                cur_num_outputs = state.op.metrics.num_outputs_generated
+                cur_time = time.time()
+                last_num_outputs, last_time = self._last_num_outputs_and_time[state.op]
+                if cur_num_outputs > last_num_outputs:
+                    self._last_num_outputs_and_time[state.op] = (
+                        cur_num_outputs,
+                        cur_time,
+                    )
+                else:
+                    if cur_time - last_time > self.MAX_OUTPUT_IDLE_SECONDS:
+                        downstream_idle = True
+                        self._print_warning(state.op, cur_time - last_time)
         return max_blocks_to_read_per_op
+
+    def _print_warning(self, op: "PhysicalOperator", no_output_time: float):
+        if self._warning_printed:
+            return
+        self._warning_printed = True
+        msg = (
+            "Temporarily unblocking backpressure."
+            f" Because some tasks of operator {op} have been submitted,"
+            f" but no outputs are generated for {no_output_time} seconds."
+            " Ignore this warning if your UDF is expected to be slow."
+            " This may also be because some resources are preempted by"
+            " non-Ray-Data workloads."
+            " If this is the case, set `ExecutionOptions.reserved_resources`."
+            " This message will only be printed once."
```

Comment:
Can we make the error message expose fewer implementation details? I don't think users will necessarily know what "backpressure" means, for example.

Maybe something like:

"Operator {op} is running, but has not produced outputs for {no_output_time}s. Ignore this warning if your UDF is expected to be slow.

Otherwise, this can happen when there are fewer cluster resources available to Ray Data than expected. If you have non-Data tasks or actors running in the cluster, reserve resources for them with `ray.data.ExecutionOptions.reserved_resources = {"num_cpus": <CPUs to exclude>}`."